### ISSUE TITLE
deluge, qbittorrent etc. "app launchers" dynamic

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#10]
they are essentially modules, and instead of copy & paste should follow dynamic structure with pre-established structure, like modules & plugins.

Not copy & pasted.

This allows more dynamic handling of additional software features, and works as the basis of supporting much more features.

**Comment by MagnaCapax** [comment]
## Adversarial Review

**Issue:** #10 — deluge, qbittorrent etc. "app launchers" dynamic
**Reviewed:** 2026-05-20
**Verdict:** NEEDS-INVESTIGATION

### Adversarial Findings

**Finding 1 (Critical): The issue has no implementation specification — "dynamic structure with modules & plugins" is not a design.** The body is two sentences. It does not name the module boundary, the data contract, the caller interface, the file layout, or the migration path for the ~10 integration subsystems each app currently spans. No implementer can start from this. A well-formed refactor issue names what the abstraction IS (not just that one should exist), what the data shape is, and what a before/after diff looks like for adding one new app.

**Finding 2 (High): The three skel `www/` toggle pages are already thin — the copy-paste problem is real but minor there; the deep duplication is elsewhere and harder.** `etc/skel/www/deluge.php` and `etc/skel/www/qbittorrent.php` are each under 30 lines and both delegate to the already-shared `pmssFrontendToggleAction()` helper in `etc/skel/.scriptsInc.php:27`. The toggle abstraction already exists. What remains per-app in those files: the enable-file path (`.delugeEnable` vs `.qbittorrentEnable`), the start-function body (binary name, log path, flags), and the kill command. These are inherently app-specific parameters — they cannot be eliminated, only data-driven. The cron watchdog layer (`scripts/cron/checkDelugeInstances.php`, `scripts/cron/checkQbittorrentInstances.php`) is already fully unified via `pmssUserWatchdogRunService()` in `userLifecycle.php`. The remaining copy-paste surface is: port validation helpers in `scripts/lib/user/torrentPort.php` (~80% shared structure between `pmssDelugePortEnsure()` and `pmssQbittorrentPortEnsure()`), and per-app sections in `etc/skel/www/welcome.php` (lines 334–399, each app ~20 lines of HTML+PHP). These are the realistic targets.

**Finding 3 (High): ADR 0016 (accepted 2026-05-17, inviolable) imposes a hard constraint the issue predates and does not account for.** Customer-facing PHP in `etc/skel/www/` must not cross-include from `/scripts/`. Any "dynamic single file" implementation that consolidates app definitions must itself live entirely within the customer tree — it cannot load app metadata from operator-side scripts. The existing `etc/skel/www/deluge.php` already had its `/scripts/` require removed per ADR 0016. A unified app registry file would need to be self-contained or served via the operator-write / customer-read artifact pattern from ADR 0016 §Option C. This is a non-trivial constraint: it means the "single file with definitions for all apps" either (a) must embed all app logic in `etc/skel/www/` with no operator-tree access, or (b) must read a data artifact written by an operator cron. The issue mentions neither path.

**Finding 4 (Medium): The issue's own TODO comments have drifted from the issue body.** The three `#TODO` references in `etc/skel/www/deluge.php:12`, `qbittorrent.php:15`, and `rclone.php:8` call for "a single file with definitions for all apps, deluge, qbittorrent, jellyfin, *ARR etc." This is a narrower, more specific ask (unified skel page) than the issue title/body ("dynamic structure with pre-established structure, like modules & plugins"), which reads as a broader architectural refactor. These are two different scopes: a data-driven skel consolidation vs. a full plugin system. A plugin system is significantly more complex and harder to maintain in a 15-year-old legacy codebase where stability is the primary constraint.

**Finding 5 (Medium): No success metric is definable from the issue as written.** What observable signal within 30 days proves this worked? The issue gives none. "Code is less copy-pasted" is not observable independently of the code's own existence. A valid metric would be: "adding a new app requires changes to N files instead of M files" with N and M stated. Without this, neither implementer nor reviewer can verify completion.

**Finding 6 (Low): The installer layer (`scripts/lib/update/apps/deluge.php`) is 388 lines of Debian-version-specific install logic, source builds, and patches — not parameterizable by a simple config entry.** Any "dynamic" installer abstraction that covers app installation (not just the skel toggle pages) must accommodate Debian 10 pip/source builds, Debian 11+ apt paths, binary symlink convergence, and Python source patching. This is not reducible to a shared data structure without significant complexity increase. Issue #10 should be scoped to the skel/welcome/port-validation surface; it should NOT be interpreted as covering the installer layer.

### Caller Analysis

`pmssFrontendToggleAction` callers (all in `etc/skel/www/`):
- `etc/skel/www/deluge.php:15`
- `etc/skel/www/qbittorrent.php:19`
- `etc/skel/www/rclone.php:11`

Distribution list in `scripts/lib/update/users/filesystem.php:187-206`: `www/deluge.php`, `www/qbittorrent.php` are both in the propagation list. Any new unified file must also be added here or it will not reach existing home directories on update.

`pmssUserWatchdogRunService` — already generic. Cron scripts are already one call each; no refactor needed there.

### Investigation Results

- **Affected files (skel toggle layer):** `etc/skel/www/deluge.php` (27 LOC), `etc/skel/www/qbittorrent.php` (31 LOC), `etc/skel/www/rclone.php` (20 LOC), `etc/skel/.scriptsInc.php` (44 LOC)
- **Affected files (port validation):** `scripts/lib/user/torrentPort.php:56-126` — `pmssDelugePortEnsure()` and `pmssQbittorrentPortEnsure()` share ~80% structure; extractable to a generic `pmssTorrentPortEnsure(string $app, ...)` with app-specific config-read callbacks
- **Affected files (welcome page):** `etc/skel/www/welcome.php:334-399` — per-app sections for Deluge, rclone, qBittorrent; each ~20 lines; data-driven approach reduces welcome.php maintenance overhead
- **Affected files (cron watchdog):** Already unified — `scripts/cron/checkDelugeInstances.php` and `scripts/cron/checkQbittorrentInstances.php` each delegate entirely to `pmssUserWatchdogRunService()`; no refactor needed
- **Call chain (skel):** `etc/skel/www/deluge.php` → `pmssFrontendToggleAction()` in `.scriptsInc.php:27` → `$startHandler()` (local function)
- **Call chain (port):** cron `checkDelugeInstances.php` → `pmssUserWatchdogRunService()` → `pmssDelugeApplyUploadThrottle()` in `scripts/lib/user/deluge.php` | `etc/skel/www/deluge.php` (skel-side) → `pmssDelugePortEnsureCurrentUser()` via `function_exists()` guard
- **Test coverage:** `scripts/lib/tests/development/torrentPortFrontendTest.php` — tests verify `pmssFrontendToggleAction()` is present, ADR 0016 compliance (`require_once '/scripts/'` absent), and port helper wiring. Any implementation touching these files must keep these tests green.
- **Git context:** Watchdog was already refactored to generic form (commits `76ff5112`, `6ad65d25`, `378691da`, `605749d8`). ADR 0016 accepted 2026-05-17 (commit `dccd3855`); ADR 0017 accepted same day. These are recent inviolable constraints.
- **Constraints:** ADR 0016 (inviolable) — no user-facing skel PHP may `require_once` from `/scripts/`. ADR 0017 — all PRs touching `etc/skel/www/*.php` require a 6-point checklist. Any implementation of this issue touching those files is subject to both ADRs.

### Recommendations

Before implementation can proceed, the following need answers:

1. **Scope decision:** Is this issue about (a) unifying the three skel toggle pages into a data-driven single file, (b) extracting shared port-validation logic in `torrentPort.php`, (c) making `welcome.php` per-app sections data-driven, or (d) all of the above? These are three independent refactors with different risk profiles. Conflating them produces an unshippable PR.

2. **ADR 0016 compliance path:** A unified "app definitions" file in `etc/skel/www/` cannot import from `/scripts/`. The app registry either lives entirely in `etc/skel/www/` (customer tree) or communicates via an operator-written data artifact. Which pattern? The choice affects the architecture of any implementation.

3. **Distribution list:** If the implementation produces a new file in `etc/skel/www/` (e.g., `www/appRegistry.php`), it must be added to the `$files` array in `scripts/lib/update/users/filesystem.php` and go through the ADR 0017 checklist. Factor this into any PR.

4. **Installer layer:** The three `etc/skel/update/apps/` installers (`deluge.php`, `rclone.php`, etc.) contain Debian-version-specific build and patch logic that is not abstractable by a shared data structure. If this issue intends to unify the installer layer, that is a separate, larger effort and should be a separate issue.

5. **Success metric:** State concretely what "done" looks like. Suggested form: "Adding a new app requires changes to N files and M lines instead of the current X files and Y lines, verified by a test that walks through adding a stub app."

### Worth-Implementing Analysis

- **Cost:** Build cost for skel-layer-only unification: low (100–200 LOC across 3–4 files). Port-validation extraction: low–medium (50 LOC refactor with test update). Welcome.php data-driven refactor: medium (careful HTML/PHP restructure). Full plugin system as implied by "modules & plugins": high (500+ LOC, new abstraction layer, migration of all existing apps).
- **Benefit:** Concrete and real — the issue was identified as the prerequisite for adding new apps without multiplying copy-paste debt. Any new app currently requires ~10 touch-points. A unified skel toggle layer reduces that to 1 new data entry + 1 new cron file. The port-validation extraction removes a clear DRY violation.
- **Maintenance cost:** A data-driven app registry requires keeping the registry in sync with installed apps; drift between registry and actual installed binaries is a new failure mode that must be considered.
- **Verdict rationale:** The refactor is worth doing in the scoped form (skel toggle + port validation + welcome.php). It is not worth doing as a full plugin system. The issue as written cannot be implemented because it specifies neither the interface nor the scope.
- **Alternative if left as-is:** Continue adding new apps by copy-paste (the Transmission issue demonstrates the cost). Acceptable only if new apps are rare; not acceptable if new apps are a growth vector.

### Anti-Pattern Hunt

- **REGEX_FOR_SEMANTICS:** PASS — issue proposes a structural refactor, no semantic classification logic involved
- **PREMATURE_OPTIMIZATION:** PASS — refactor addresses a real, growing debt demonstrated by the ~10-file integration surface per new app
- **OPERATOR_HANDOFF_ANTIPATTERN:** PASS — issue does not propose delegating to operator for routine operations
- **LAZY_HALT_AFTER_AUTHORIZATION:** PASS — not applicable to this refactor issue
- **FALSE_COMPLETION_CLAIM:** PASS — issue is a feature request, not a completion claim
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — no invented paths or authorities
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS — existing `function_exists()` guards in skel pages already fail closed gracefully
- **DRY_RUN_THEATER:** PASS — not applicable to this issue type
- **INVESTIGATION_LOOP_THEATER:** PASS — this is a first-time review
- **LACK_OF_INVESTIGATION:** FAIL — the issue body contains no investigation of the actual copy-paste surface, no enumeration of the affected subsystems, no proposed interface, and no scope boundary. An implementer reading only the issue has no actionable starting point. The issue is a valid observation but not a workable specification.

### Decomposition Targets

If the issue is reworked into implementable scope, the natural decomposition is three independent PRs, each self-contained:

1. **Port validation DRY extraction** (`scripts/lib/user/torrentPort.php:56-126`): Extract shared structure of `pmssDelugePortEnsure()` and `pmssQbittorrentPortEnsure()` into a generic `pmssTorrentPortEnsure(string $appName, string $portFile, callable $configReader, callable $configWriter, string $processBinary, ...)`. Low blast radius; operator-only code; existing tests cover behavior.

2. **Welcome.php per-app sections data-driven** (`etc/skel/www/welcome.php:334-399`): Replace three copy-pasted binary-detect + button-render blocks with a loop over an app definition array. Each entry: binary path(s), enable-file, action-endpoint, button labels, tab hint text. ADR 0016 and 0017 apply; render harness must pass.

3. **Skel toggle pages unified** (`etc/skel/www/deluge.php`, `qbittorrent.php`, `rclone.php`): If and only if a genuine new app is being added (providing the forcing function), unify the three ~25-LOC files into a single `appLauncher.php` that reads app identity from a URL parameter or a per-app include. Distribution list in `filesystem.php` must be updated; ADR 0017 checklist applies; existing `TorrentPortFrontendTest` must stay green.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review (Update)

**Issue:** #10 — deluge, qbittorrent etc. "app launchers" dynamic
**Reviewed:** 2026-05-27
**Verdict:** NEEDS-INVESTIGATION (unchanged from prior review 2026-05-20)

> **Context:** A prior adversarial review was posted 2026-05-20. This update addresses new findings from codebase changes and related issues that appeared after that review. The prior review's core analysis (no spec, three independent scopes, ADR 0016 constraint, five questions for the operator) remains accurate and is not repeated in full here.

---

### New Findings Since 2026-05-20

**[HIGH] Finding 7: The `function_exists()` port-enforce guards in `deluge.php` and `qbittorrent.php` are permanently dead code.**

Both skel files contain:

```php
if (function_exists('pmssDelugePortEnsureCurrentUser')) {
    pmssDelugePortEnsureCurrentUser();
}
```

`pmssDelugePortEnsureCurrentUser()` is defined in `scripts/lib/user/torrentPort.php`. Per ADR 0016 (accepted 2026-05-17, inviolable), customer-facing PHP cannot `require` anything from `/scripts/` — the directory is `750 root:root`, user PHP CGI cannot traverse it. The `function_exists()` guard will always return `false`. The function is never loaded in the customer context.

Issue #574 ("torrentPort.php: no operator cron calls port-ensure functions — enforcement currently inactive") confirms this explicitly: port enforcement is inactive fleet-wide. Any refactor of these files under this issue should **remove these dead guards** rather than preserve them. Their presence misleads future readers into believing enforcement runs on the customer-call path. It does not.

The correct call path for port enforcement is an operator-side cron (see #574 for design options). This is a separate fix — but any implementation of this issue that touches `deluge.php` or `qbittorrent.php` should remove the dead guards as a first-principles cleanup.

**[MEDIUM] Finding 8: `qbittorrent.php` carries a stale joke comment that is misleading in isolation.**

`etc/skel/www/qbittorrent.php` line ~23:
```php
function startQbittorrent() {    // this actually calls the function to start rTorrent :)
```

The comment reads as "this function starts rTorrent, not qBittorrent" — the opposite of what the function body does (`qbittorrent-nox -d`). Whether this is a historical artifact of copy-paste from the rTorrent handler or an in-joke that has lost context, it is confusing to any implementer touching this file. Remove in any refactor touching this file.

**[MEDIUM] Finding 9: ADR 0017 (accepted 2026-05-17, binding) adds a 6-point checklist gate to every PR touching `etc/skel/www/*.php` — this was not reflected in the prior review's scope assessment.**

Any implementation of this issue that modifies or creates files under `etc/skel/www/` must satisfy:

1. Require-path safety (no `/scripts/` paths — CI-enforced)
2. Function-call safety: every called function must be defined in the customer tree or be a PHP built-in — CI-enforced via `scripts/testing/customer-context-fatal-scan.php`
3. Render safety: renders cleanly under simulated customer context — CI-enforced via `scripts/testing/customer-panel-render-harness.php`
4. Architectural fit: no admin/write logic mixed into customer files
5. Distribution-list registration: any new file must be added to the `$files` array in `scripts/lib/update/users/filesystem.php`
6. Test coverage: unit or characterization test must use the customer-tree path

This checklist means the prior review's decomposition targets now carry explicit CI enforcement cost. The implementer cannot rely on passing existing tests alone — tests 2 and 3 above specifically target the failure class (#508: fatal 500 from unreachable `/scripts/` require) that triggered ADR 0016.

**[LOW] Finding 10: The `welcome.php` per-app sections (prior review Decomposition Target 2) have been significantly refactored by commits since 2026-05-20.**

Seven commits since the prior review have restructured the welcome panel. The per-app sections (Deluge, rclone, qBittorrent) are now at lines 317-337 and already use `pmssWelcomeServiceAvailable()` and `pmssWelcomeActionButtonHtmlBuild()` helper functions — 3-5 lines per app, not the "~20 lines each" from the prior review. The data-driven loop consolidation described in Decomposition Target 2 remains possible but the duplication is already substantially reduced. This decomposition target has lower ROI than stated in the prior review.

---

### Updated Call Chain Analysis

The prior review's call chain for port enforcement:
> `etc/skel/www/deluge.php` (skel-side) → `pmssDelugePortEnsureCurrentUser()` via `function_exists()` guard

**This call chain is dead.** The `function_exists()` guard never fires in the customer context. The prior review described it as active with an `is_readable()` guard; that guard was removed by commit `4732b070` (2026-05-24) because it too was always-false. The customer-side call path for port enforcement does not exist.

Surviving call chain for port validation (operator-side only):
- `scripts/lib/user/torrentPort.php` — `pmssDelugePortEnsure()`, `pmssQbittorrentPortEnsure()` — defined, tested, **no caller in production** (issue #574)

---

### Related Issues Cross-Reference

Any implementation of this issue should coordinate with:

- **#508** (open, customer-reported): documents the concrete customer harm of the prior `/scripts/` require violation. The bare `require_once` has been removed from current skel files; the issue appears resolved by the ADR 0016 compliance work but has not been formally closed.
- **#574** (open, blocked): port enforcement inactive fleet-wide. The `function_exists()` dead guards in `deluge.php` and `qbittorrent.php` track back to this. Resolve before implementing port-enforcement aspects of any new "dynamic app" structure.
- **#511** (open): asserts propagation-list membership for all `www/` frontends. Any new file added to `etc/skel/www/` by this issue should satisfy #511's coverage intent.

---

### Recommendations (additions to prior review)

6. **Remove dead `function_exists()` guards** from `deluge.php` and `qbittorrent.php` as a prerequisite cleanup, or include their removal as part of any implementation PR. Dead code that looks like active enforcement is worse than no code.
7. **Apply ADR 0017 checklist in the implementation PR** — all 6 points, not just the CI-caught ones. Points 2 and 3 require active verification by the implementing agent.
8. **Do not implement port-enforcement logic in the skel layer** — route that to #574 (operator cron) and keep the skel layer purely presentation.

### Worth-Implementing Analysis (unchanged)

Assessment from 2026-05-20 stands: scoped skel-layer unification is worth doing. Full plugin architecture is not. The issue still requires operator scope decision before implementation can start. No new evidence changes this.

### Anti-Pattern Hunt (delta from prior review)

All prior findings hold. One update:

- **INVESTIGATION_LOOP_THEATER:** PASS — this is a second-pass update with genuine new information (dead code confirmed, ADR 0017 codified, new related issues filed); not a stale loop.

### Decomposition Targets (updated)

**[REFACTOR — priority raised]** `etc/skel/www/deluge.php:22-24` and `etc/skel/www/qbittorrent.php:27-29` — remove the permanently-dead `function_exists('pmssDelugePortEnsureCurrentUser')` and `function_exists('pmssQbittorrentPortEnsureCurrentUser')` guards. Zero behavior change; removes misleading dead code that suggests active port enforcement where none exists.

**[REFACTOR — priority lowered]** Welcome.php data-driven loop (prior Decomposition Target 2) — welcome.php refactoring since the prior review has reduced the per-app section size from ~20 LOC to ~4 LOC. Remaining consolidation into a data-driven loop is still valid but lower urgency than stated in the prior review.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Investigation Report

**Issue:** #10 — deluge, qbittorrent etc. "app launchers" dynamic
**Investigated:** 2026-05-29
**Status:** PARTIAL

### Summary
Third pass on #10 (after the 2026-05-20 and 2026-05-27 reviews). The key update is that **the code moved since the last review**: the dead `function_exists()` port guards and the stale comment flagged previously are gone, and the launcher/management surfaces are now genuinely modular. The "pre-established structure like modules" the issue asks for already exists at every layer that can sensibly carry one. What remains is either inherently per-app (installer build logic) or a speculative full "plugin registry" that is YAGNI. Recommendation: **SHOULD-NOT-IMPLEMENT** the plugin registry; the actionable de-duplication has substantially landed.

### What The Issue Asks For
App launchers should follow a "dynamic structure with pre-established structure, like modules & plugins", not copy-paste, to serve as "the basis of supporting much more features." No interface, data contract, file layout, or scope boundary is specified — the core gap both prior reviews identified.

### Implementation Status
The module pattern is **already present** at each launcher/management layer (verified against current source):

- **Skel toggle pages** — `etc/skel/www/deluge.php` (12 LOC), `qbittorrent.php` (13 LOC), `rclone.php` (12 LOC) are now pure per-app config passed to the shared `pmssFrontendToggleAction(enableFile, startHandler, disableCommand, restartCommand)` helper in `etc/skel/.scriptsInc.php:143`.
- **Dead guards removed** — commit `45fce340` (2026-05-27) removed the permanently-false `function_exists('pmssDelugePortEnsureCurrentUser')` guards from `deluge.php`/`qbittorrent.php` (they could never fire under ADR 0016). `grep function_exists` on the three files now returns nothing. The stale `startQbittorrent()` comment is also gone. Findings 7 & 8 of the 2026-05-27 review are resolved.
- **Welcome panel** — `etc/skel/www/welcome.php:331/342/348` uses `pmssWelcomeServiceAvailable('<app>.php', [binaries])` + `pmssWelcomeActionButtonHtmlBuild(...)`; ~4-6 LOC per app.
- **Watchdog crons** — `scripts/cron/checkDelugeInstances.php` (24 LOC) and `checkQbittorrentInstances.php` (17 LOC) are pure config calls to `pmssUserWatchdogRunService(...)`.
- **Torrent counting** — `scripts/userTorrents.php:24-29` is already a per-app glob-pattern array.
- **Port read** — common parts already extracted to `pmssTorrentPortExpectedRead()` and `pmssTorrentPortCurrentUserContext()` in `scripts/lib/user/torrentPort.php`.
- **Distribution list** — `scripts/lib/update/users/filesystem.php` registers `.scriptsInc.php`, `www/deluge.php`, `www/qbittorrent.php`, `www/welcome.php`. Note: `www/rclone.php` is absent here — propagation concern tracked under #511, not this issue.

### What's Missing
Surfaces that remain per-app — but are **inherently** so, or legacy/minor (not eliminable by a data structure):

- **Installer** `scripts/lib/update/apps/<app>.php` — ~388-409 LOC of Debian-version-specific build/patch/symlink logic per client. Not abstractable without a large complexity increase. Out of scope (both prior reviews agree).
- **User library** `scripts/lib/user/<app>.php` and **dpkg baseline** — app-specific behavior / package sets.
- **Port-ensure remainder** in `torrentPort.php` — `pmssDelugePortEnsure()` (web.conf, JSON-ish) vs `pmssQbittorrentPortEnsure()` (INI) handle genuinely different config formats; the "~80% shared" figure from prior reviews is overstated now that the common read/context helpers are extracted.
- **stats.php:412-424** app-status map is a small `ps`-grep table (and currently omits qbittorrent — a minor pre-existing gap, separate concern).
- **nginx** `template.nginx-user` per-app blocks are **legacy-compat only** (`/deluge-##username/`, kept until 2028); the canonical `/user-##username/<app>/` path is generic, so a new app on the canonical path needs no per-app nginx.

The maximal "plugin registry" that auto-wires all surfaces from one definition is **not built** — and should not be (see Recommendation).

### Devil's Advocate
*Why NOT build the dynamic/plugin structure?*
1. The launcher surfaces that CAN be data-driven already are; the registry's marginal value is small.
2. The remaining surfaces are irreducibly per-app — a registry relocates that code, it does not delete it.
3. No forcing function: no new torrent client is imminent (Transmission #31 was verdicted should-not-implement; Docker-rootless is non-viable on the VM topology per #171). Building the abstraction speculatively risks the *wrong* abstraction.
4. Roadmap "more features" (Invidious #52, remote desktop #57, S3 #40, LinuxServer.io #18) are Docker/different-architecture and would not fit a torrent-client-shaped registry.

*Rebuttal / what would change this:* a concrete decision to add several new torrent-client-style apps would supply the forcing function — at which point the residual launcher-layer duplication should be folded into that work. A registry is feasible within ADR 0016/0017 (a self-contained skel file, or a data artifact written operator-side and read by customer PHP); it is feasible-but-not-recommended, not impossible.

### Architectural Minimization
Applying delete-first:
- **DELETE** — the dead `function_exists()` guards: already deleted (`45fce340`). The "plugin registry" non-feature is best deleted by not building it; the per-layer helpers already provide the structure. **This is the winning path.**
- **MERGE** — welcome per-app sections into one loop; converge stats.php app-status onto the same definition. Candidates exist but are low-ROI and unforced.
- **REFACTOR** — a generic `pmssTorrentPortEnsure()` over the two port-ensure functions: rejected — the config formats differ enough that callback indirection would exceed the ~10 truly-shared lines, and it depends on #574.
- **NULL** — if nothing is done, adding a (currently non-existent) new torrent client costs a few small config entries plus the inherent installer/library work: acceptable.

### ADR compatibility
The recommendation (do not add a cross-tree registry) is the maximally compliant outcome under ADR 0016 (customer-tree PHP must not `require` from `/scripts/`) and ADR 0017 (the 6-point `etc/skel/www/*.php` checklist). Any future registry remains feasible only as a self-contained skel file or an operator-written / customer-read artifact under both ADRs.

### Anti-Pattern Hunt
- **REGEX_FOR_SEMANTICS:** PASS — structural refactor; no semantic-classification logic.
- **PREMATURE_OPTIMIZATION:** FAIL for the maximal plugin-registry reading (does not exist; not forced; no current consumer; costly now) — which is exactly why the recommendation is SHOULD-NOT-IMPLEMENT. The already-landed per-layer helpers are NOT premature (they address realized duplication).
- **FALSE_COMPLETION_CLAIM:** PASS — status is explicitly PARTIAL; done vs remaining surfaces are reconciled with file:line evidence.
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — every path/commit/ADR verified from source and git log.
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS — toggle endpoints fail closed (an explicit `action` is required via `pmssFrontendActionRequest()`, which `die()`s otherwise).
- **DRY_RUN_THEATER:** PASS — not applicable.

### Related Context
The 2026-03-28 review of Transmission (#31) documented that adding a new torrent client then spanned ~10 integration surfaces and named #10 as the architectural prerequisite — true at that time, when only the watchdog was modular. Refactoring since has modularized the toggle, welcome, watchdog, counting, and port-read layers, largely satisfying that prerequisite for the launcher/management surfaces. No customer-facing incident is attributable to #10 itself; the related panel-render failure class (unreachable `/scripts/` requires, ADR 0016) is tracked under #508.

### Cross-Issue Analysis
- **#316** (sustained refactoring; eliminate `function_exists` guards) — the umbrella under which the #10 launcher de-duplication and guard removal actually happened; same "No Copy-Pasta" principle. Residual micro-DRY folds here. Adjacent, not a clean supersession.
- **#574** (port-ensure functions have no operator-cron caller; enforcement inactive) — **DEPENDENCY**: any port-ensure refactor touches the same file; #574 (functional) should land before #10's cosmetic port-DRY.
- **#470** (per-user `rtorrentEnabled` toggle) — adjacent per-user app-toggle pattern; not a duplicate.
- **#559** (move *arr management out of system `update.php`) — adjacent app-management architecture (opposite direction: per-account separation); not a duplicate.
- **#23** ("Deluge plugins") — name-similar but concerns Deluge's own plugin ecosystem; unrelated to a PMSS app framework.

### Dependencies & Blockers
- ADR 0016 (inviolable): customer-tree PHP must not `require` from `/scripts/`. Any registry must be self-contained in the customer tree or read an operator-written artifact.
- ADR 0017 (binding): 6-point checklist on every `etc/skel/www/*.php` PR.
- #574 must land before any port-ensure consolidation.

### Reuse, Refactor & Test Strategy
- **Existing reuse points:** `pmssFrontendToggleAction()` (`.scriptsInc.php:143`), `pmssWelcomeServiceAvailable()` / `pmssWelcomeActionButtonHtmlBuild()` (`welcome.php`), `pmssUserWatchdogRunService()` (cron watchdogs), `pmssTorrentPortExpectedRead()` / `pmssTorrentPortCurrentUserContext()` (`torrentPort.php`), the per-app glob array (`userTorrents.php`).
- **Refactor opportunities (only under a forcing function):** fold welcome per-app sections into a single loop over an app-definition array; converge stats.php app-status onto the same definition. Both low-ROI standalone.
- **Test categories:** any skel-touching change must keep `scripts/lib/tests/development/torrentPortFrontendTest.php` green and satisfy the ADR 0017 CI layers (`customer-php-tree-isolation`, `customer-context-fatal-scan`, `customer-panel-render-harness`). Regression target: the ADR 0016 empty-panel class.
- **Minimum-code-for-correctness:** no new code is warranted now; the smallest correct change set is empty until a new app provides the forcing function.

### Recommendation
**SHOULD-NOT-IMPLEMENT** (the speculative plugin registry).

The de-duplication this issue called for has substantially landed through sustained refactoring (#316 umbrella): the launcher toggle pages, welcome panel, watchdog crons, torrent counting, and port-read helpers are modular, and the dead guards are gone. The remaining per-app surfaces are inherently app-specific (installer/library/dpkg) or legacy/minor. A full plugin registry is YAGNI and risks the wrong abstraction; it should be revisited only under a concrete new-app forcing function, folding residual launcher-layer DRY in at that point. The port-DRY remnant additionally depends on #574.

This is a technical verdict, not a deferral. The close decision on this issue is left to the maintainer's discretion.

**Suggested ADR** — title: *"App-launcher modularization scope: per-layer shared helpers, no speculative plugin registry."* It would record that PMSS app launchers are modularized via per-layer shared helpers; that a unified plugin registry is deferred as YAGNI because the remaining surfaces are inherently per-app and no new-app forcing function exists; and that a registry remains feasible within ADR 0016/0017 if a future forcing function appears. Recording this prevents a fourth re-investigation of #10 and gives future contributors the integration-surface map for the forcing-function case.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review (Verdict)

**Issue:** #10 — deluge, qbittorrent etc. "app launchers" dynamic
**Reviewed:** 2026-05-29
**Verdict:** SHOULD-NOT-IMPLEMENT

> **Context:** Today's Investigation Report (same date, comment above) covers the full surface-by-surface codebase analysis. This review formalises the verdict with the canonical format, incorporates current codebase verification, and applies the merit-check framework.

---

### Adversarial Findings

**[Critical] No concrete present demand.** The issue was filed 2026-02-01. No new torrent-client-style app is currently in flight: Transmission (#31, SHOULD-NOT-IMPLEMENT), Docker-rootless approaches (non-viable on the VM topology per #171). The forcing function that would justify building a plugin registry does not exist. Without it, any registry built now is speculation against an unknown future app shape.

**[High] The copy-paste problem has substantially been solved by the per-layer helpers that have landed since this issue was filed.** Verified against current source: `etc/skel/www/deluge.php` (14 LOC), `qbittorrent.php` (13 LOC), `rclone.php` (14 LOC) are pure config calls to `pmssFrontendToggleAction()` in `etc/skel/.scriptsInc.php`. Watchdog crons delegate entirely to `pmssUserWatchdogRunService()` with per-app specs. Welcome panel uses `pmssWelcomeServiceAvailable()` + `pmssWelcomeActionButtonHtmlBuild()` (~4–6 LOC per app). Port-read helpers (`pmssTorrentPortExpectedRead()`, `pmssTorrentPortCurrentUserContext()`) are already extracted. The issue is largely resolved by the existing refactoring campaign.

**[Medium] Building the registry now risks the wrong abstraction.** `welcome.php` app sections, watchdog specs, and toggle pages are each ~4–14 LOC. The remaining per-app surfaces (installers, user libraries, dpkg sets) are inherently app-specific — a registry cannot eliminate them, only relocate them. Speculative unification produces a registry that does not match the actual anatomy of the next app added, requiring retcon under ADR 0016/0017 constraints.

**[Low — Watchdog operator-side note]** The watchdog crons use `function_exists('pmssDelugeApplyUploadThrottle')` / `function_exists('pmssQbittorrentApplyUploadThrottle')` guards. Unlike the now-deleted customer-side guards, these are operator-side and active: the watchdog framework loads the app library (last argument to `pmssUserWatchdogRunService()`), so the function exists when the guard fires. Not dead code; no action needed.

---

### Caller Analysis

No removal of functionality is proposed by this issue. The existing per-layer helpers are retained regardless of verdict.

---

### Worth-Implementing Analysis — Merit Check (Tier A)

**1. Concrete present demand.** Theoretical — no present demand. The issue body states a general principle; no ticket, customer complaint, or blocked feature depends on a plugin registry today.

**2. Function.** Create a shared app-definition registry in `etc/skel/www/` that data-drives the toggle pages, welcome panel entries, and watchdog cron specs so adding a new torrent-client-style app requires editing one registry file rather than N per-layer files.

**3. Success metric.** "Adding a new torrent-client-style app requires changes to 1 registry file instead of the current ~5 per-layer files, verified by walking a stub app through the install flow." The metric is not observable today because no new app is being added — which is itself the decisive SHOULD-NOT-IMPLEMENT signal per the merit check framework.

**4. Catastrophic if implemented.** Registry shape chosen now → first real new app has genuinely different requirements (different binary lifecycle, no web UI, different port scheme) → registry does not fit → registry is patched with special cases → ADR 0016/0017 compliance cost per per-skel-file edit → registry code becomes the complexity it was meant to prevent. Detection lag: not visible until the new app is added, potentially months out. Recovery: `git revert` is clean but means redoing the app integration without the registry.

**5. Catastrophic if NOT implemented.** Nothing observable — the per-layer helpers already exist. If a new torrent-client-style app is added, ~5 files get new entries of ~10–15 LOC each. No customer harm, no service disruption, no security gap. This asymmetry with #4 is the verdict signal.

**6. Deletion alternative.** Not building the registry IS the deletion path. The abstraction should emerge from use (second new app) rather than being invented ahead of it.

**7. Cost.** (a) Build: 200–400 LOC (registry file + migration of 3 apps + ADR 0017 checklist × 3+ files). (b) Maintenance: registry must stay in sync with installed binaries; drift produces confusing customer-panel buttons — a new failure mode absent from the per-layer-helper model. (c) Blast radius if broken: cosmetic (wrong buttons shown), low severity. (d) Undo: clean `git revert`.

**8. Dependency blast.** `pmssFrontendToggleAction()` callers: `deluge.php`, `qbittorrent.php`, `rclone.php`. Distribution list in `scripts/lib/update/users/filesystem.php`. Watchdog crons. All light; no external consumers outside PMSS.

**9. Requirements challenge.** The dumbest implicit requirement is "the abstraction should be built before any app needs it" — because YAGNI exists precisely to prevent this. The correct requirement is "the abstraction should be built when adding the second new app, after confirming the first new app's shape on real data."

**Tier B — Gall + Second-System (speculative v2 of working system).** The per-layer helpers ARE the working v1. A unified registry is a proposed v2 being built speculatively. The incremental path being skipped: add an actual new app → identify duplication in that real app → refactor with real data. Scope smuggled in beyond stated claim: installer abstraction (too complex, explicitly out of scope in prior reviews), user-library abstraction (inherently per-app), nginx per-app legacy blocks (not needed for canonical-path apps).

**Verdict:** SHOULD-NOT-IMPLEMENT. No present demand, no observable success metric, catastrophic-if-NOT is "nothing observable," deletion alternative (not building) is the preferred path. The issue's intent is substantially satisfied by the existing per-layer refactoring.

---

### Anti-Pattern Hunt

- **REGEX_FOR_SEMANTICS:** PASS — structural refactor, no semantic classification logic.
- **PREMATURE_OPTIMIZATION:** FAIL (plugin-registry reading) — registry does not exist, no current consumer, measured cost of not having it is low. All four proofs fail. Root cause of SHOULD-NOT-IMPLEMENT, not a separate disqualifier.
- **OPERATOR_HANDOFF_ANTIPATTERN:** PASS — issue does not push execution work to operator.
- **LAZY_HALT_AFTER_AUTHORIZATION:** PASS — not applicable.
- **FALSE_COMPLETION_CLAIM:** PASS — today's Investigation Report (same date) explicitly labels status PARTIAL; no false completion.
- **FABRICATED_AUTHORITY_OR_PATH:** PASS — all cited files, ADRs (0016, 0017), and commit hashes verified from source and `git log`.
- **FAIL_OPEN_SAFETY_BEHAVIOR:** PASS — toggle endpoints require explicit `action` parameter; `pmssFrontendActionRequest()` calls `die()` without it.
- **DRY_RUN_THEATER:** PASS — not applicable.
- **INVESTIGATION_LOOP_THEATER:** PASS — this closes a three-review investigation arc (2026-05-20, 2026-05-27, today) with a terminal verdict.
- **LACK_OF_INVESTIGATION:** PASS — counter-signals checked: (1) source files read directly; (2) git log reviewed for post-Investigation-Report commits (`80cb64d6` watchdog safety improvement — does not affect verdict); (3) `torrentPort.php` helper extraction verified at line level; (4) related issues #574, #511, #508 cross-referenced.

---

### Decomposition Targets

No new targets beyond those in today's Investigation Report. The port-ensure DRY opportunity (two separate `PortEnsure` functions in `torrentPort.php`) remains contingent on #574. Welcome.php per-app loop consolidation remains valid but low-ROI and unforced.

---

*Note for the operator: The Investigation Report (same date) recommends filing an ADR recording that PMSS app launchers are modularised via per-layer shared helpers and that a unified plugin registry is deferred pending a concrete new-app forcing function. This would prevent repeat investigation of this question. Labels updated: `should-not-implement` added, `needs-investigation` removed.*

— Väinämöinen
